### DISCUSSION TITLE
How do I combine multiple PDF files into a single document

### DISCUSSION THREAD

**Discussion by IngoManthey** [QuestPDF/QuestPDF#1074] [Q&A] [answered]
For example, I have created 10 individual invoices. I would like to summarize them in one document. Unfortunately, I cannot find any information about this in the help.
My problem is how can I load a PDF and then add the other 9?  List<string> fileNames;

```
DocumentOperation
    .LoadFile("document1.pdf")
    .MergeFile("document2.pdf")
    .Save("merged.pdf");

```
Or do I have to call this function 8 times??

So it's pointless to save the document 8 times.

**Comment by girlpunk** [comment]
As `MergeFile` returns a `DocumentOperation` instance, you should be able to call `MergeFile` again to add additional files. For example:
```csharp
DocumentOperation
    .LoadFile("document1.pdf")
    .MergeFile("document2.pdf")
    .MergeFile("document3.pdf")
    .MergeFile("document4.pdf")
    .Save("merged.pdf");
```

**Comment by IngoManthey** [comment]
Thanks for the help. But I don't have a constant name and number of invoices. If I have created 100 invoices, the file names are saved in a list of strings.
What should the code look like?
DocumentOperation
    .LoadFile("document1.pdf")
Foreach(var item in filenames)
 .MergeFile(item)

Do no work

**Comment by IIZard** [comment]
Edit: I realised you needed to load the first doc, and then merge the rest in order. I'll leave the checks for if there's less than 2 documents to you :) I also realised I did some bits wrong, hopefully it's more accurate. 

I suggest trying to write an extension method that exploits recursion to achieve what you want.
I had to use a similar pattern here:
https://github.com/QuestPDF/QuestPDF/discussions/1065

But you might be able to try something like:

```
public static DocumentOperation MergeDocuments(this DocumentOperation op, IEnumerable<string> docs)
{
    var docstomerge = docs
        //First is Loaded, not merged
        .Skip(1)
        .GetEnumerator();

    return op
        .LoadFile(doc.First())
        .MergeNext(docstomerge);
}

private static DocumentOperation MergeNext(this DocumentOperation op, IEnumerator<string> docs)
{
    //Keep a ref to the current doc for when we .MoveNext()
    var thisdoc = docs.Current;

    //If there's a next file, merge the current one and recurse the next file, otherwise just merge.
    return docs.MoveNext() ? op.MergeFile(thisdoc, null).MergeNext(docs) : op.MergeFile(thisdoc, null);
}

```

I've not tested this code so please give it a go and let me know if it works?

**Reply by IIZard** [reply]
@IngoManthey I updated the code to be more accurate, please let me know if that's better?

**Comment by IngoManthey** [comment, marked as answer]
@IIZard 
```
public static class DocumentOperationExtensions
{
    public static DocumentOperation MergeDocuments(this DocumentOperation op, IEnumerable<string> docs)
    {
        var docstomerge = docs
            //First is Loaded, not merged
            .Skip(1)
            .GetEnumerator();

        return op
            .LoadFile(doc.First())
            .MergeNext(docstomerge);
    }

    private static DocumentOperation MergeNext(this DocumentOperation op, IEnumerator<string> docs)
    {
        //Keep a ref to the current doc for when we .MoveNext()
        var thisdoc = docs.Current;

        //If there's a next file, merge the current one and recurse the next file, otherwise just merge.
        return docs.MoveNext() ? op.MergeFile(thisdoc, null).MergeNext(docs) : op.MergeFile(thisdoc, null);
    }
}

```
I have pasted your code into my project. Unfortunately I get the following compiler error with: .LoadFile(doc.First()):
Compiler error CS0176
Article
09/15/2021
8 contributors
Static member 'member' cannot be accessed with an instance reference; qualify it with a type name instead

Only a class name can be used to qualify a static variable; an instance name cannot be a qualifier. For more information, see Static Classes and Static Class Members.

I will try to load the first file outside.
Something is wrong. I have my extension class in the same namespace.

```
 DocumentOperation.LoadFile(fileNames[0]);
 DocumentOperation.MergeDocuments(fileNames);
```
But I get the message
DocumentOperation does not contain a definition for MergeDocuments

```
var firstName = fileNames[0];
fileNames.Remove(firstName);
var op = DocumentOperation.LoadFile(firstName);
var saveName = GetFileName("Invoices.pdf", PathNames.MainTemp);
foreach (var item in fileNames)
{
    op.MergeFile(item);
}
op.Save(saveName);
```
This is how it works well

**Comment by DawidRobertNowak** [comment]
Hello,

Glad you were able to find the solution! We are closing this thread as resolved. Thank you for sharing your answer — it may help others with the same question.

For reference, the full documentation on merging and other document operations can be found here:
https://www.questpdf.com/concepts/document-operations.html

We are also considering adding an example for merging a dynamic number of files, as this is a common use case that is not currently covered in the documentation.

Thank you!
